srml/executive/src/lib.rs

```diff
@@ -265,10 +262,13 @@ impl<
 			<system::Module<System>>::note_extrinsic(encoded);
 		}
 
+		if let Some(d) = xt.doughnut() {
+			storage::unhashed::put(well_known_keys::DOUGHNUT_KEY, &d);
+		}
+
 		// Decode parameters and dispatch
 		let (f, s) = xt.deconstruct();
 		let r = f.dispatch(s.into());
-		storage::unhashed::kill(well_known_keys::DOUGHNUT_KEY);
```

Comment:
Let's remove the doughnut from storage post-storage.
Reasoning: Without removing it, and if the next extrinsic to execute does not have a doughnut, it will use the previous extrinsic's doughnut.

Reply:
More robust: add the removal before the `put` logic above (line 264.
Reasoning: This handles all edge-cases besides storage failure.